src/app-layer-ftp.c

```diff
@@ -228,31 +272,121 @@ static int FTPParseRequest(Flow *f, void *ftp_state,
     /* toserver stream */
     state->direction = 0;
 
+    int direction = STREAM_TOSERVER;
     while (FTPGetLine(state) >= 0) {
         FTPParseRequestCommand(state,
                                state->current_line, state->current_line_len);
-        if (state->command == FTP_COMMAND_PORT) {
-            if (state->current_line_len > state->port_line_size) {
-                ptmp = SCRealloc(state->port_line, state->current_line_len);
-                if (ptmp == NULL) {
-                    SCFree(state->port_line);
-                    state->port_line = NULL;
-                    state->port_line_size = 0;
-                    return 0;
+        switch (state->command) {
+            case FTP_COMMAND_PORT:
+                if (state->current_line_len > state->port_line_size) {
+                    ptmp = SCRealloc(state->port_line, state->current_line_len);
+                    if (ptmp == NULL) {
+                        SCFree(state->port_line);
+                        state->port_line = NULL;
+                        state->port_line_size = 0;
+                        return 0;
+                    }
+                    state->port_line = ptmp;
+
+                    state->port_line_size = state->current_line_len;
                 }
-                state->port_line = ptmp;
-
-                state->port_line_size = state->current_line_len;
-            }
-            memcpy(state->port_line, state->current_line,
-                   state->current_line_len);
-            state->port_line_len = state->current_line_len;
+                memcpy(state->port_line, state->current_line,
+                        state->current_line_len);
+                state->port_line_len = state->current_line_len;
+                break;
+            case FTP_COMMAND_RETR:
+                /* change direction (default to server) so expectation will handle
+                 * the correct message when expectation will match.
+                 */ 
+                direction = STREAM_TOCLIENT;
+                // fallthrough
+            case FTP_COMMAND_STOR:
+                {
+                    struct FtpTransferCmd *data = SCCalloc(1, sizeof(struct FtpTransferCmd));
+                    if (data == NULL)
+                        SCReturnInt(-1);
+                    data->DFree = FtpTransferCmdFree;
+                    data->filename = SCCalloc(state->current_line_len - 4, sizeof(char));
+                    if (data->filename == NULL) {
+                        SCFree(data);
+                        SCReturnInt(-1);
+                    }
+                    memcpy(data->filename, state->current_line + 5, state->current_line_len - 5);
```

Comment:
it seems the goal here is to create a nul-terminated string. Two things:

1. please make this explicit
2. is this guaranteed to be a valid string? As it comes from the wire my suspicion is 'no'. So a byte array instead?

Reply:
From https://tools.ietf.org/html/rfc959 at section 5.3.2 we are suppose to have ASCII only here. I'm adding a check here. Mister @chifflier do you know if I should be more laxist here ?

Reply:
Making it a byte array makes it match behavior of our other protocol parsers and gives consistent matching logic in our keywords. So this is what I think we should do.

Reply:
OK, will do that.